services/src/main/java/org/keycloak/services/DefaultKeycloakSession.java

```diff
@@ -337,8 +338,19 @@ public UserCredentialManager userCredentialManager() {
     }
 
     @Override
-    @SuppressWarnings("unchecked")
     public <T extends Provider> T getComponentProvider(Class<T> clazz, String componentId) {
+        final RealmModel realm = getContext().getRealm();
+        if (realm == null) {
+            throw new IllegalArgumentException("Realm not set in the context.");
+        }
+
+        // Loads componentModel from the realm
+        return this.getComponentProvider(clazz, componentId, KeycloakModelUtils.componentModelGetter(realm.getId(), componentId));
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public <T extends Provider> T getComponentProvider(Class<T> clazz, String componentId, Function<KeycloakSessionFactory, ComponentModel> modelGetter) {
         Integer hash = clazz.hashCode() + componentId.hashCode();
```

Comment:
@tnorimat I am not sure if your concern is valid, but in theory it may be.

However this part is not something, which is introduced in this PR. Note that PR just introduced `getComponentProvider` with the `modelGetter` as an argument, but this algorithm of computing hashes was already there in the original `getComponentProvider` method. Also same "issue" is in method `getProvider(Class<T> clazz, String id)` and as far as I remember, it works like this for years and I never recall any conflict. I think your concern might be valid, but it is likely out of scope of this PR.

But anyway, I suggest if you send an email to keycloak-dev with this concern of computing hashes and potential conflict related to it?

Reply:
@mposolda IMO, there is very little chance that the current codes cause conflicting hash values. Therefore, we need not take this comment as an issue. Sorry for bothering you.

Reply:
No problem. I am myself not 100% sure if this concern is valid, but I suppose it can be. So if you prefer, the email to keycloak-dev regarding this concern can be fine. But just if you prefer to do it...

Reply:
@mposolda I've re-considered this comment and concluded there is no problem.

Consider
```
 H = hash(A) + hash(B)
```

I was afraid that accidentally A' and B' are created that satisfy the following and such several pairs of A' and B' exist. 

```
H = hash(A') + hash(B')
```

However, consider
```
 H' := H - hash(A') 
 H' = hash(B')
```
which means that the probability of arising this situation is equal to that of colliding a single integer's hash value with other integer's hash value.

If H consists of sum of N hash values of an integer (N >=2),  it follows the statement just above.

Reply:
@tnorimat Cool, I think you're right and the chance of conflict is close to zero. Thanks for the doublecheck